### ISSUE TITLE
Force Stepwise to Explore All Possibilities

### ISSUE THREAD

**Issue by AlexanderLavelle** [alkaline-ml/pmdarima#438]
Hello - 
My boss uses R but my peer group uses Python. We are doing ts forecasting and using pmdarima for AR/MA ordering. 

Is there a way to force the stepwise search to exhaust available options before finishing? It appears to use some sort of gradient but sometimes does not persist to the higher orders with lower AIC. 

I also utilize StepwiseContext(max_iterations=100), max_order >= 20, and max_iter=100, but the search rarely utilizes the entire grid and may not force the best models. 

This behavior apparently is different than R, where I am told the behavior exhausts the grid. 

Do you have any recommendations to expand the search OR to behave as in R?

Thank you in advance!

--------------------------------------------------------

System:
    python: 3.8.5 (default, Sep  3 2020, 21:29:08) [MSC v.1916 64 bit (AMD64)]
executable: C:\ProgramFiles\Anaconda3\python.exe
   machine: Windows-10-10.0.18362-SP0

Python dependencies:
        pip: 21.0.1
 setuptools: 52.0.0.post20210125
    sklearn: 0.24.2
statsmodels: 0.12.2
      numpy: 1.20.1
      scipy: 1.6.2
     Cython: 0.29.23
     pandas: 1.2.4
     joblib: 1.0.1
   pmdarima: 1.8.2
Windows-10-10.0.18362-SP0
Python 3.8.5 (default, Sep  3 2020, 21:29:08) [MSC v.1916 64 bit (AMD64)]
pmdarima 1.8.2
NumPy 1.20.1
SciPy 1.6.2
Scikit-Learn 0.24.2
Statsmodels 0.12.2

**Comment by tgsmith61591** [comment]
Hi Alexander,

There is a misconception here -- the stepwise algorithm is _not_ exhaustive (in either R or `pmdarima`). It follows the step-wise algorithm laid out in section 2.2 of Rob Hyndman's 2008 paper: ["Automatic Time Series Forecasting"](https://www.researchgate.net/publication/222105759_Automatic_Time_Series_Forecasting_The_forecast_Package_for_R).

If you want to perform a sweep of the entire parameter space (honoring `max_p`, `max_q`, `max_P`, `max_Q` and `max_order`), you can achieve this without the stepwise algorithm, but it's generally slower:

```python
In [1]: import pmdarima as pm

In [2]: y = pm.datasets.load_wineind()

In [3]: model = pm.auto_arima(y, stepwise=False, trace=3)
 ARIMA(0,1,0)(0,0,0)[1] intercept   : AIC=3587.776, Time=0.03 sec
 ARIMA(0,1,1)(0,0,0)[1] intercept   : AIC=3509.903, Time=0.07 sec
 ARIMA(0,1,2)(0,0,0)[1] intercept   : AIC=3506.259, Time=0.12 sec
 ARIMA(0,1,3)(0,0,0)[1] intercept   : AIC=3506.173, Time=0.10 sec
 ARIMA(0,1,4)(0,0,0)[1] intercept   : AIC=3503.947, Time=0.16 sec
 ARIMA(0,1,5)(0,0,0)[1] intercept   : AIC=3495.968, Time=0.24 sec
 ARIMA(1,1,0)(0,0,0)[1] intercept   : AIC=3573.940, Time=0.01 sec
 ARIMA(1,1,1)(0,0,0)[1] intercept   : AIC=3508.854, Time=0.14 sec
 ARIMA(1,1,2)(0,0,0)[1] intercept   : AIC=3505.978, Time=0.17 sec
Near non-invertible roots for order (1, 1, 3)(0, 0, 0, 1); setting score to inf (at least one inverse root too close to the border of the unit circle: 1.000)
 ARIMA(1,1,3)(0,0,0)[1] intercept   : AIC=inf, Time=0.23 sec
 ARIMA(1,1,4)(0,0,0)[1] intercept   : AIC=3504.179, Time=0.26 sec
 ARIMA(2,1,0)(0,0,0)[1] intercept   : AIC=3540.400, Time=0.03 sec
 ARIMA(2,1,1)(0,0,0)[1] intercept   : AIC=3500.610, Time=0.10 sec
 ARIMA(2,1,2)(0,0,0)[1] intercept   : AIC=3503.442, Time=0.16 sec
Near non-invertible roots for order (2, 1, 3)(0, 0, 0, 1); setting score to inf (at least one inverse root too close to the border of the unit circle: 0.996)
 ARIMA(2,1,3)(0,0,0)[1] intercept   : AIC=inf, Time=0.28 sec
 ARIMA(3,1,0)(0,0,0)[1] intercept   : AIC=3516.118, Time=0.03 sec
 ARIMA(3,1,1)(0,0,0)[1] intercept   : AIC=3502.598, Time=0.15 sec
 ARIMA(3,1,2)(0,0,0)[1] intercept   : AIC=3504.888, Time=0.08 sec
 ARIMA(4,1,0)(0,0,0)[1] intercept   : AIC=3516.442, Time=0.03 sec
 ARIMA(4,1,1)(0,0,0)[1] intercept   : AIC=3501.754, Time=0.19 sec
 ARIMA(5,1,0)(0,0,0)[1] intercept   : AIC=3517.051, Time=0.04 sec

Best model:  ARIMA(0,1,5)(0,0,0)[1] intercept
Total fit time: 2.634 seconds
```

For more info, take a look at [the documentation](http://alkaline-ml.com/pmdarima/modules/generated/pmdarima.arima.auto_arima.html#pmdarima.arima.auto_arima) for the `auto_arima` method.
